config/initializers/routing_hash_for.rb

```diff
@@ -2,26 +2,27 @@ module ActionDispatch
   module Routing
     class RouteSet
       class NamedRouteCollection
-        def define_url_helper(route, name, options)
-          helper = UrlHelper.create(route, options.dup)
-
-          @module.remove_possible_method name
-          @module.module_eval do
+        def define_url_helper(mod, route, name, opts, route_key, url_strategy)
+          helper = UrlHelper.create(route, opts, route_key, url_strategy)
+          mod.module_eval do
             define_method(name) do |*args|
-              helper.call self, args
+              options = nil
+              options = args.pop if args.last.is_a? Hash
+              helper.call self, args, options
             end
 
             #because we heavily rely on the removed hash_for method in routes, we must add this monkey patch.
             define_method("hash_for_#{name}") do |*args|
-              unless args.first.is_a? Hash
-                id = args.first.try(:to_param)
-                options[:id] = id if id.present?
-              end
-              helper.send(:handle_positional_args, self, args, options, [])
+              inner_options = nil
+              inner_options = args.pop if args.last.is_a? Hash
+              helper.send(:handle_positional_args,
+                          {},
+                          inner_options || {},
+                          args,
+                          opts.dup,
```

Comment:
The argument is modified inside `handle_positional_args` when it's merged with the other parameters passed into the hash_to_* URL helper: https://github.com/rails/rails/blob/v4.2.6/actionpack/lib/action_dispatch/routing/route_set.rb#L305

Note that this doesn't affect the regular url helper, only our `hash_for_` extension, because the hash is duplicated inside the UrlHelper.call method here: https://github.com/rails/rails/blob/v4.2.6/actionpack/lib/action_dispatch/routing/route_set.rb#L273

If we don't #dup it here, there will be some integration test failures when running the full suite, or running the provisioning templates functionals followed by the audit integration tests in the same process.  This happens after the provisioning template history tab is rendered, which links to audits using `hash_for_audits_path` ([source](https://github.com/theforeman/foreman/blob/1.11.0-RC2/app/helpers/audits_helper.rb#L81)).

This causes `opts` (which is an empty hash of parameters for this route, from config/routes.rb) to get `:search => 'user => something'` permanently stored in it, which then affects the audits integration test as _every_ audits_path link in the app inherits the same search parameter!  The same bug affected roles integration tests too.